### PR TITLE
Upgrade esp-hal to 0.18.0

### PR DESCRIPTION
**Pull request by nickfarrow**
Compiles using the cst816s fork @musdom found.

Also upgrades to embedded_hal 1.0.

Currently the screen display is blank.

* The new `mipsi::Builder` wants to receive an `SpiDevice` and not the `SpiBus`, so we need to choose a CS pin? Not sure what to put here.
* Is `Orientation::Portrait(false)` equal to `Rotation:Deg90`?
* I'm also setting the reset pin to high (https://github.com/almindor/mipidsi/blob/master/docs/TROUBLESHOOTING.md)

### CODE REVIEW

**Comment by musdom** [comment]
Display works but now top devices upstream go in uart mode

**Comment by musdom** [comment]
Todo: remove tick constant in esp32_run.rs and io.rs

**Comment by musdom** [comment]
Rebased over #120

File: Cargo.toml [hunk 1]
```diff
@@ -11,8 +11,15 @@ resolver = "2"
...
-serde = { version = "1", features = ["derive","alloc"], default-features = false }
+bincode = { version = "2.0.0-rc.3", features = [
+    "serde",
+    "alloc",
+    "derive",
+], default-features = false }
+serde = { version = "1", features = [
+    "derive",
+    "alloc",
+], default-features = false }
```
Comment (LLFourn) [inline Cargo.toml:22]: Can we not have this weird formatting please? I'm guessing this was done automatically without noticiing?

File: Cargo.toml [hunk 2]
```diff
@@ -26,3 +33,18 @@ sha2 = { version = "0.10", default-features = false }
 # bdk_chain = { path = "../bdk/crates/chain" }
 bdk_chain = { git = "https://github.com/LLFourn/bdk.git", rev = "165f06b8f78bd24f57dfbfd66da1faa189c67457" }
 bdk_electrum = { git = "https://github.com/LLFourn/bdk.git", rev = "165f06b8f78bd24f57dfbfd66da1faa189c67457" }
+
+[profile.dev]
+# Rust debug is too slow.
+# For debug builds always builds with some optimization
+opt-level = "s"
```
Comment (LLFourn) [inline Cargo.toml:40]: I think this is a good idea. I've been just doing --release even when testing because of this.

File: Cargo.toml (hunk 1 above)
Comment (musdom) [reply Cargo.toml:22]: github CI formatted it because it wasn't like this in my local repo

File: Cargo.toml [hunk 3]
```diff
@@ -26,3 +33,18 @@ sha2 = { version = "0.10", default-features = false }
...
+
+[profile.release]
+# LLVM can perform better optimizations using a single thread
+codegen-units = 1
+debug = 2
+debug-assertions = false
+incremental = false
+lto = 'fat'
+opt-level = 's'
```
Comment (LLFourn) [inline Cargo.toml:49]: Why are we doing this in the workspace Cargo.toml and not the individual crates. Surely we'll want a different profile for device than everything else.

File: device/Cargo.toml [hunk 4]
```diff
@@ -16,23 +16,26 @@ mem_debug = []
...
+display-interface-spi = { version = "0.5.0", optional = true }
+mipidsi = { version = "0.8.0", features = ["batch"], optional = true }
 embedded-graphics-framebuf = { version = "0.5.0", optional = true }
 embedded-graphics = "0.8.1"
 embedded-text = "0.7.0"
 u8g2-fonts = { version = "0.3.0", features = ["embedded_graphics_textstyle"] }
 rand_chacha = { workspace = true }
-cst816s = "0.1.4"
-embedded-hal = { version = "0.2.7" }
+cst816s = { git = "https://github.com/fxweidinger/cst816s", rev = "82f88186e94a94126b962d30b4c0129fd3c1c5f8" }
```
Comment (LLFourn) [inline device/Cargo.toml:35]: We should probably vendor this inside the workspace at some point if we're going to use it in production.

File: Cargo.toml (hunk 3 above)
Comment (musdom) [reply Cargo.toml:49]: Cargo complains profiles should be in the root Cargo.toml

Comment (LLFourn) [reply Cargo.toml:49]: Apparently not possible so I guess we'll just suck it up for now but at some point we may want to split out device into its own repo.

File: device/src/bin/v2.rs [hunk 5]
```diff
@@ -109,13 +112,14 @@ fn main() -> ! {
...
         .with_mosi(io.pins.gpio7);
-    let di = SPIInterfaceNoCS::new(spi, io.pins.gpio9.into_push_pull_output());
-    let display = Builder::st7789(di)
-        .with_display_size(240, 280)
-        .with_window_offset_handler(|_| (0, 20)) // 240*280 panel
-        .with_invert_colors(ColorInversion::Inverted)
-        .with_orientation(Orientation::Portrait(false))
-        .init(&mut delay, Some(io.pins.gpio6.into_push_pull_output())) // RES
+    let spi_device = embedded_hal_bus::spi::ExclusiveDevice::new_no_delay(spi, NoCs);
+    let di = SPIInterface::new(spi_device, Output::new(io.pins.gpio9, Level::Low));
```
Comment (LLFourn) [inline device/src/bin/v2.rs:116]: It'd be interesting to know what the `embedded_hal_bus` crate is doing that the original Spi type was not. It sounds like from the documentation it helps different devices share the bus. Are we doing that?

Comment (musdom) [reply device/src/bin/v2.rs:116]: Nope, it's just for the display

Comment (LLFourn) [reply device/src/bin/v2.rs:116]: Right so why do we need the `embedded_hal_bus` crate. What goes wrong without it?

**Review by LLFourn — APPROVED** [review]
Working nicely for me and LGTM.

File: device/src/bin/v2.rs (hunk 5 above)
Comment (musdom) [reply device/src/bin/v2.rs:116]: display_interface_spi expects that SpiDevice embedded-hal-bus type now

**Pull request merged**